pincer/commands/chat_command_handler.py

```diff
@@ -0,0 +1,435 @@
+    increase lookup time.
+    The way this problem is avoided is by storing a version of the commands that we can
+    deal with as library developers and a version of the command that Discord thinks we
+    should provide. That is where the register and the built_register help simplify the
+    design of the library.
+    The register is simply where the "Pincer version" of commands gets saved to memory.
+    The built_register is where the version of commands that Discord requires is saved.
+    The register allows for O(1) lookups by storing commands in a Python dictionary. It
+    does cost some memory to save two copies in the current iteration of the system but
+    we should be able to drop the built_register in runtime if we want to. I don't feel
+    that lost maintainability from this is optimal. We can index by in O(1) by checking
+    the register but can still use the built_register if we need to do a nested lookup.
+
+    Attributes
+    ----------
+    client: :class:`Client`
+        The client object
+    managers: Dict[:class:`str`, :class:`~typing.Any`]
+        Dictionary of managers
+    register: Dict[:class:`str`, :class:`~pincer.objects.app.command.InteractableStructure`]
+        Dictionary of ``InteractableStructure``
+    built_register: Dict[:class:`str`, :class:`~pincer.objects.app.command.InteractableStructure`]
+        Dictionary of ``InteractableStructure`` where the commands are converted to
+        the format that Discord expects for sub commands and sub command groups.
+    """  # noqa: E501
+
+    has_been_initialized = False
+    managers: List[Interactable] = []
+    register: Dict[str, InteractableStructure] = {}
+    built_register: Dict[str, AppCommand] = {}
+
+    # Endpoints:
+    __get = "/commands"
+    __delete = "/commands/{command.id}"
+    __update = "/commands/{command.id}"
+    __add = "/commands"
+    __add_guild = "/guilds/{command.guild_id}/commands"
+    __get_guild = "/guilds/{guild_id}/commands"
+    __update_guild = "/guilds/{command.guild_id}/commands/{command.id}"
+    __delete_guild = "/guilds/{command.guild_id}/commands/{command.id}"
+
+    def __init__(self, client: Client):
+        self.client = client
+        self._api_commands: List[AppCommand] = []
+        _log.debug(
+            "%i commands registered.", len(ChatCommandHandler.register.items())
+        )
+
+        self.__prefix = f"applications/{self.client.bot.id}"
+
+    async def get_commands(self) -> List[AppCommand]:
+        """|coro|
+
+        Get a list of app commands from Discord
+
+        Returns
+        -------
+        List[:class:`~pincer.objects.app.command.AppCommand`]
+            List of commands.
+        """
+        # TODO: Update if discord adds bulk get guild commands
+        guild_commands = await gather(*(
+            self.client.http.get(
+                self.__prefix + self.__get_guild.format(
+                    guild_id=guild.id if isinstance(guild, Guild) else guild
```

Comment:
What if we take this step to preload and cache all the guilds?

Reply:
not a problem for this PR